### ISSUE TITLE
[Feature Request]: State performance improvements

### ISSUE THREAD

**Issue by damccorm** [apache/beam#22964]
### What would you like to happen?

As follow up to #22736 there are a few things we can do to improve state performance. These are:

1. Caching of state
2. Deferred state reads (readLater and prefetched) - https://beam.apache.org/documentation/programming-guide/#deferred-state-reads

Caching is a larger dig, but readLater and prefetched should be relatively straightforward - they just require running a Get request in parallel with the main computation and adding some thread safety to exec/userstate.go. Prefetch will also require some graph parsing to know when it should perform an early read (see https://docs.google.com/document/d/1rcKa1Z6orDDFr1l8t6NA1eLl6zanQbYAEiAqk39NQUU/edit?disco=AAAAdca3izg for details on how you could do this).

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-go
